nemo_aligner/models/mm/stable_diffusion/megatron_sd_draftp_model.py

```diff
@@ -0,0 +1,381 @@
+            )
+
+            # # transforms trajectores from a list of T+1 (b x image) tensors to a tensor of shape (b * (T+1), ...)
+            trajectories_predx0 = (
+                torch.stack(intermediates["pred_x0"], dim=0)
+                .transpose(0, 1)
+                .contiguous()
+                .view(-1, *intermediates["pred_x0"][0].shape[1:])
+            )
+
+            vae_decoder_output = []
+            idx_denoised_imgs = [self.inference_steps + (self.inference_steps + 1) * i for i in range(batch_size)]
+
+            for i in range(0, batch_size, self.vae_batch_size):
+                image = self.model.differentiable_decode_first_stage(
+                    trajectories_predx0[idx_denoised_imgs[i : i + self.vae_batch_size]]
+                )
+
+                vae_decoder_output.append(image)
+
+            vae_decoder_output = torch.cat(vae_decoder_output, dim=0)
+            vae_decoder_output = torch.clip((vae_decoder_output + 1) / 2, 0, 1) * 255.0
+
+            log_reward = [
+                self.reward_model.get_reward(
+                    vae_decoder_output[i].unsqueeze(0).detach().permute(0, 2, 3, 1), batch  # (C, H, W) -> (1, H, W, C)
+                ).item()
+                for i in range(batch_size)
+            ]
+            log_img = [
+                np.transpose(vae_decoder_output[i].float().detach().cpu().numpy(), (1, 2, 0))  # (C, H, W) -> (H, W, C)
+                for i in range(batch_size)
+            ]
+            return log_img, log_reward
+
+    @torch.no_grad()
+    def log_visualization(self, prompts):
+
+        batch_size = len(prompts)
+        # Get different seeds for different dp rank
+        with get_cuda_rng_tracker().fork(get_data_parallel_rng_tracker_name()):
+            latents = torch.randn(
+                [
+                    batch_size,
+                    self.in_channels,
+                    self.height // self.downsampling_factor,
+                    self.width // self.downsampling_factor,
+                ],
+                generator=None,
+            ).to(torch.cuda.current_device())
+
+        image_draft_p, reward_draft_p = self.generate_log_images(latents, prompts, self.model)
+        image_init, reward_init = self.generate_log_images(latents, prompts, self.init_model)
+
+        images = []
+        captions = []
+        for i in range(len(image_draft_p)):
+            images.append(image_draft_p[i])
+            images.append(image_init[i])
+            captions.append("DRaFT+: " + prompts[i] + ", Reward = " + str(reward_draft_p[i]))
+            captions.append("SD: " + prompts[i] + ", Reward = " + str(reward_init[i]))
+
+        self.wandb_logger.loggers[1].log_image(
+            key="Inference Images",
+            images=[wandb.Image(Image.fromarray(img.round().astype("uint8"))) for img in images],
+            caption=captions,
+        )
+
+    def generate(
+        self, batch, x_T,
+    ):
+
+        with torch.cuda.amp.autocast(
```

Comment:
we talked last time about disabling this autocast if amp_o2 =True, did you check if that's possible?

Reply:
Yeah Ming connected me to mlperf team and here is their reply:

"I'll need to go through the codebase. Generally, even with AMP O2, we need to cast the model params to FP16, make a copy of params in FP32 for master params and so forth. I am not sure if this can be removed unless the flag amp_O2 takes care of everything under the hood."

I think it's safer to keep it for now

Reply:
O2 does take care of this for LLMs, but if this is what we need to do it's fine for now